views: access control

nharraud opened the pull request:
* NEW Adds customizable access control to record views. Allow
  configuring different permissions per endpoint. (reference #14)

Signed-off-by: Nicolas Harraudeau <nicolas.harraudeau@cern.ch>

jirikuncar commented on invenio_records_rest/ext.py:
```diff
@@ -26,9 +26,66 @@
+        self._update_permission_factory = None
+        self._delete_permission_factory = None
+
+    @property
```
IMHO it can be `werkzeug.utils:cached_property`

jirikuncar commented on invenio_records_rest/views.py:
```diff
@@ -42,10 +42,29 @@
 from .serializers import record_to_json_serializer
 
 
+current_read_permission_factory = LocalProxy(
```
isn't it enough to have just `current_records_rest`?

jirikuncar commented on invenio_records_rest/views.py:
```diff
@@ -194,6 +272,17 @@ def get(self, pid, record, **kwargs):
         :param record: Record object.
         :returns: The requested record.
         """
+        # Check permissions
```
perfect candidate for a decorator

jirikuncar commented on tests/test_invenio_records_rest.py:
```diff
@@ -90,6 +90,7 @@ def control_num(data, cn=1):
 
 def test_valid_create(app, resolver):
     """Test VALID record creation request (POST .../records/)."""
+    InvenioRecordsREST(app)
```
IMHO it should be part of `app` fixture

lnielsen commented on examples/permsapp.py:
```diff
@@ -0,0 +1,170 @@
+    $ curl -XPOST http://localhost:5000/login -d 'email=admin%40invenio-software.org&password=123456' -c mycookie
+    $ curl -XGET http://localhost:5000/records/1 -b mycookie
+
+Login as forbidden@invenio-software.org (password 123456), hwo has not
```
hwo -> who

lnielsen commented on examples/permsapp.py:
```diff
@@ -0,0 +1,170 @@
+InvenioAccess(app)
+InvenioRecordsREST(app)
+
+rec_uuid = '9107e6ef-fea4-4971-ae2f-934d2fdcaa30'
```
Probably we should use the same uuid as in records-ui:
``deadbeef-1234-5678-ba11-b100dc0ffee5``

lnielsen commented on invenio_records_rest/ext.py:
```diff
@@ -44,10 +101,25 @@ def init_app(self, app):
+            "RECORDS_REST_DEFAULT_DELETE_PERMISSION_FACTORY",
+            "invenio_records.permissions:delete_permission_factory")
+
+        app.config.setdefault("RECORDS_REST_LOGIN_ENDPOINT", "security.login")
```
Do we redirect in the rest API? shouldn't we just return a 401 error.

lnielsen replied:
Yes, it does return 401 error, so I guess this one isn't needed?

nharraud replied:
Yes you are right, bad copy paste from invenio-records-ui...

nharraud commented:
@lnielsen @jirikuncar One thing you can see in the code but which I forgot to mention is that the tests and example use the UI login as we don't have oauth yet. As the search-ui will be also use cookie authentication to access this REST API this seemed a valid usage even if not the most common.

jirikuncar commented:
@nharraud both OAuthClient and OAuth2Server should be ready on GitHub (not PyPI yet).

nharraud commented:
@jirikuncar Sorry I meant Application Token. There is no need for an OAuth authentication here. Using an application token added directly to the database would remove the dependency on UI login.

nharraud commented:
Coverage decreased because I added the delete permission factory methods even though there is no delete method implemented yet.

nharraud commented:
@jirikuncar can we release a new version of invenio-rest? This depends on the last commits.

nharraud commented:
@jirikuncar I removed the example with permissions, removed default permission factories and copied permissions.py from invenio_records for the tests.

jirikuncar commented on invenio_records_rest/views.py:
```diff
@@ -178,15 +266,52 @@ class RecordResource(ContentNegotiatedMethodView):
+        try:
+            record.delete()
+            # mark all PIDs as DELETED
+            all_pids = PersistentIdentifier.query \
```
IMHO you can use only one call of `.filter()` (or `db.and_(cond1, cond2)`).
```python
all_pids = PersistentIdentifier.query.filter(
    PersistentIdentifier.object_type == pid.object_type,
    PersistentIdentifier.object_uuid == pid.object_uuid
).all()
```

jirikuncar commented on tests/test_invenio_records_rest.py:
```diff
@@ -81,6 +84,18 @@ def create_record(data):
 
+def delete_record(pid, record):
+    with db.session.begin_nested():
+        record.delete()
```
We should consider hooking PIDStore marking as deleted to record signal.

--
cc @lnielsen

nharraud replied:
@jirikuncar I am not sure doing the deletion automatically is the best way to do it. We would automate only part of the process. The pid creation is not automatic either. The records-rest API is creating both the pid and the record separately. It will also be complicated to detect the merging and mark the pid as *moved*.
Also the pidstore module does not depend on invenio-records (or only for the tests).

nharraud replied:
@jirikuncar If it is only a concern on code duplication I will call the delete from the REST API instead. It will be cleaner that way.

jirikuncar commented on tests/permissions.py:
```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015, 2016 CERN.
```
if it's a new file it should contain only current year

jirikuncar commented on .gitignore:
```diff
@@ -55,3 +55,6 @@ docs/_build/
 target/
+
+# Sqlite databases
+*.db
```
the better solution would be to do proper cleaning after running tests

nharraud replied:
In fact these are not for tests, which are cleaning properly. This is for the examples. The `app.py` example has its instance directory inside the `examples` directory for two reasons. The first one is that the directory was not created if it did not exist. The second reason is that every module's example is named `app,py` and thus has the same instance director. I had to clean it manually each time in order to be sure that the database was in a clean state.
Thus I create an instance directory for each example in the examples directory.

jirikuncar replied:
Wouldn't `examples/instance/` do the job?

nharraud replied:
I had multiple examples at that time, one without access control and one with it. I have removed it so that it does not depend on the deprecated invenio_records/permissions.py but I might add another example providing its own permission_factories later.

jirikuncar commented on invenio_records_rest/ext.py:
```diff
@@ -26,9 +26,66 @@
+        """Load default read permission factory."""
+        if self._read_permission_factory is None:
+            imp = self.app.config[
+                "RECORDS_REST_DEFAULT_READ_PERMISSION_FACTORY"
```
It is not critical, but can you use single quotes for strings for consistency with rest of the code. They are randomly mixed through the code also in other files. Thanks

nharraud replied:
Yes you are right. Can I fix that in the next pull request (search integration) so that I don't spend too much time rebasing?

jirikuncar replied:
IMHO it's not so much pain to run search&replace. You won't have many conflicts when you rebase other branches.

nharraud replied:
Ok. Is there a guideline for quotes in Invenio? I think we should also fix the cookie cutter as it is putting `"` in some places.

jirikuncar replied:
Just use your judgement and be consistent.

jirikuncar commented on tests/__init__.py:
```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
```
Please remove this file as its usage is discouraged with `py.test`. 

> **avoid “__init__.py” files in your test directories.** This way your tests can run easily against an installed version of mypkg, independently from the installed package if it contains the tests or not.

*source https://pytest.org/latest/goodpractises.html*

jirikuncar commented:
- [ ] Update copyright year in **all** files you have touched in this PR. Thanks

nharraud commented:
@jirikuncar Done. Only the `.gitignore` has not been updated as I didn't see copyrights in other projects either.
